File: Project-source/forge-verifier/typescript/src/online.ts
// Online / hybrid composition — parses + verifies signature/expiry, then
// dispatches CRL refresh + heartbeat based on the license's `mode` field.
//
// Fail-closed policy mirrors the Go / Java / C# / Rust SDKs:
//   - offline → never touches the network
//   - hybrid  → signature MUST pass; CRL + heartbeat best-effort
//               (cached CRL revoke still wins)
//   - online  → signature + CRL + heartbeat ALL MUST pass; net error → reject

import type { CrlClient } from "./crl.js";
import type { HeartbeatClient } from "./heartbeat.js";
import { parse } from "./verify.js";
import { verifyFile } from "./verify.js";
import { ForgeError, type VerifyResult } from "./types.js";

export interface OnlineVerifierOptions {
  publicKey: Uint8Array;
  heartbeat?: HeartbeatClient;
  crl?: CrlClient;
}

export class OnlineVerifier {
  constructor(private readonly opts: OnlineVerifierOptions) {}

  async verify(path: string, fingerprint: string, now: Date): Promise<VerifyResult> {
    const file = await parse(path);
    const res = verifyFile(file, this.opts.publicKey, now);
    const mode = file.payload.mode ?? "offline";
    const licenseId = file.payload.license_id;

    switch (mode) {
      case "offline":
        return res;

      case "hybrid":
        if (this.opts.crl) {
          try {
            await this.opts.crl.refresh();
          } catch {
            // hybrid: ignore network errors, fall back to cached snapshot
          }
          if (this.opts.crl.isRevoked(licenseId)) {
            throw new ForgeError("license revoked (CRL)", "revoked", {
              ...res,
              status: "revoked",
            });
          }
        }
        if (this.opts.heartbeat) {
          try {
            await this.opts.heartbeat.send(licenseId, fingerprint);
          } catch {
            // best-effort
          }
        }
        return res;

      case "online":
        if (!this.opts.crl || !this.opts.heartbeat) {
          throw new ForgeError(
            "online mode requires CrlClient + HeartbeatClient",
            "malformed",
          );
        }
        try {
          await this.opts.crl.refresh();
        } catch (e) {
          throw new ForgeError(
            `online: CRL refresh failed: ${(e as Error).message}`,
            "signature_invalid",
          );
        }
        if (this.opts.crl.isRevoked(licenseId)) {
          throw new ForgeError("license revoked", "revoked", { ...res, status: "revoked" });
        }
        try {
          const hb = await this.opts.heartbeat.send(licenseId, fingerprint);
          if (hb.license_status && hb.license_status !== "active" && hb.license_status !== "valid") {
            throw new ForgeError(
              `online: server reports license_status=${hb.license_status}`,
              "signature_invalid",
            );
          }
        } catch (e) {
          if (e instanceof ForgeError) throw e;
          throw new ForgeError(
            `online: heartbeat failed: ${(e as Error).message}`,
            "signature_invalid",
          );
        }
        return res;

      default:
        throw new ForgeError(`unknown license mode: ${mode}`, "malformed");
    }
  }
}

File: Project-source/forge-verifier/rust/src/online.rs
//! Online / hybrid composition — parses + verifies signature/expiry, then dispatches
//! CRL refresh + heartbeat based on the license's own `mode` field.
//!
//! Fail-closed policy mirrors the Go / Java / C# SDKs:
//!   - `offline` → never touches network
//!   - `hybrid`  → signature MUST pass; CRL + heartbeat best-effort; cached revoke wins
//!   - `online`  → all three MUST pass; network failure → reject

use std::path::Path;

use chrono::{DateTime, Utc};

use crate::crl::CrlClient;
use crate::heartbeat::HeartbeatClient;
use crate::{parse, ForgeError, Status, VerifyResult};

pub struct OnlineVerifier<'a> {
    pub public_key: &'a [u8],
    pub heartbeat: Option<&'a HeartbeatClient>,
    pub crl: Option<&'a CrlClient>,
}

impl<'a> OnlineVerifier<'a> {
    pub fn verify(
        &self,
        path: impl AsRef<Path>,
        fingerprint: &str,
        now: DateTime<Utc>,
    ) -> Result<VerifyResult, ForgeError> {
        let parsed = parse(path)?;
        let res = parsed.verify(self.public_key, now)?;
        let license_id = parsed.payload.license_id.clone();
        let mode = parsed.payload.mode.as_str();

        match mode {
            "offline" | "" => Ok(res),

            "hybrid" => {
                if let Some(crl) = self.crl {
                    let _ = crl.refresh(); // best-effort
                    if crl.is_revoked(&license_id) {
                        let mut r = res.clone();
                        r.status = Status::Revoked;
                        return Err(ForgeError::Revoked { license_id, result: r });
                    }
                }
                if let Some(hb) = self.heartbeat {
                    let _ = hb.send(&license_id, fingerprint); // best-effort
                }
                Ok(res)
            }

            "online" => {
                let crl = self.crl.ok_or(ForgeError::OnlineRequiresClients)?;
                let hb = self.heartbeat.ok_or(ForgeError::OnlineRequiresClients)?;
                crl.refresh()
                    .map_err(|e| ForgeError::OnlineFailed(format!("CRL refresh: {e}")))?;
                if crl.is_revoked(&license_id) {
                    let mut r = res.clone();
                    r.status = Status::Revoked;
                    return Err(ForgeError::Revoked { license_id, result: r });
                }
                let hb_resp = hb
                    .send(&license_id, fingerprint)
                    .map_err(|e| ForgeError::OnlineFailed(format!("heartbeat: {e}")))?;
                if !hb_resp.license_status.is_empty() && hb_resp.license_status != "active" {
                    return Err(ForgeError::OnlineFailed(format!(
                        "server reports license_status={}",
                        hb_resp.license_status
                    )));
                }
                Ok(res)
            }

            other => Err(ForgeError::UnknownMode(other.to_string())),
        }
    }
}
